### ISSUE TITLE
Generic method inference differs between source and bytecode before JDK 17.0.19

### ISSUE THREAD

**Issue by vlsi** [uber/NullAway#1791]
### What happens

Inference for a generic method reports a failure when the method is read from a class file, and does not when the same method is read from source. The two paths agree from JDK 17.0.19 onwards; before that they disagree.

Given `NullUtil` on the classpath rather than in the compilation:

```java
package com.uber;
import org.jspecify.annotations.NullMarked;
import org.jspecify.annotations.Nullable;
@NullMarked
public class NullUtil {
  @SuppressWarnings("NullAway")
  public static <T> T assumeNonNull(@Nullable T object) {
    return object;
  }
}
```

analyzing

```java
package com.uber;
import java.util.List;
import org.jspecify.annotations.NullMarked;
import org.jspecify.annotations.Nullable;
import static com.uber.NullUtil.assumeNonNull;
import java.util.concurrent.atomic.AtomicReference;
@NullMarked
class Test {
  private final AtomicReference<@Nullable List<String>> ref = new AtomicReference<>();
  void test() {
    for (String s: assumeNonNull(ref.get())) {
      System.out.println(s);
    }
  }
}
```

reports two diagnostics that the all-source compilation does not:

```text
Test.java:11: warning: [NullAway] inference failure: type variable T is constrained to be @Nullable, but its upper bound requires it to be @NonNull
    for (String s: assumeNonNull(ref.get())) {
                                ^
Test.java:11: warning: [NullAway] enhanced-for expression 'assumeNonNull(ref.get())' is @Nullable
```

Flags: `-XepOpt:NullAway:AnnotatedPackages=com.uber`, `-XepOpt:NullAway:OnlyNullMarked=true`, plus the JSpecify mode arguments from `JSpecifyJavacConfig.withJSpecifyModeArgs`, which include `-XDaddTypeAnnotationsToSymbol=true`.

### Affected versions: fixed in 17.0.19

I bisected the JDK 17 update line, holding Error Prone fixed at 2.42.0 (the version `testJdk17` pins) and running the snippet in both modes:

| JDK | Vendor | Bytecode run |
| --- | --- | --- |
| 17.0.9+9 | IBM Semeru (OpenJ9) | fails |
| 17.0.10+13 | BellSoft Liberica | fails |
| 17.0.15+10 | BellSoft Liberica | fails |
| 17.0.17+15 | BellSoft Liberica | fails |
| 17.0.18+10 | BellSoft Liberica | fails |
| **17.0.19+11** | **BellSoft Liberica** | **passes** |
| 17.0.20+10 | BellSoft Liberica | passes |
| 17.0.20+8 | IBM Semeru (OpenJ9) | passes |
| 21, 25 | any | passes |

Two vendors move together across the boundary, so this is a fix in the OpenJDK 17 update line rather than a vendor build. The source run passes on every version above.

CI does not see this: `actions/setup-java` with `java-version: 17` installs the latest Temurin 17, which is past the boundary.

### A related observation, not investigated

On the same older JDK 17 builds, four tests unrelated to this one also fail, and pass from 17.0.19:

- `WildcardTests.issue1671`
- `BytecodeGenericsTests.genericsChecksForParamPassingAndReturns`
- `BytecodeGenericsTests.genericsChecksForFieldAssignments`
- `BytecodeGenericsTests.callMethodTakingJavaUtilFunction`

Measured on 17.0.10+13 and 17.0.18+10 (four failures) against 17.0.19+11 (none). Whether they share a cause with this issue I did not check. The practical consequence is that `./gradlew :nullaway:testJdk17` is red on a JDK 17 older than 17.0.19 and green from it, which is worth knowing before chasing a local failure.

### Where this is recorded

`GenericMethodTests.issue1453` runs in both modes as of #1789, which runs every snippet test against bytecode as well as source. The bytecode variant is gated on the boundary above:

```java
Assume.assumeTrue(
    "the bytecode run fails before JDK 17.0.19; see https://github.com/uber/NullAway/issues/1791",
    testMode == TestMode.SOURCE
        || Runtime.version().compareToIgnoreOptional(Runtime.Version.parse("17.0.19")) >= 0);
```

so every JDK that has the fix keeps running it and will fail if the behavior regresses there.

### Reproducing

Removing the assumption and running `:nullaway:testJdk17` on a JDK older than 17.0.19 reproduces it. Two things make this easy to measure wrongly, and both cost me a run:

- Gradle picks the toolchain itself, and `-Porg.gradle.java.installations.paths` did not override it here. Setting `Test.executable` to `<jdk>/bin/java` does.
- With `org.gradle.caching=true` the `Test` task's cache key does not include the JDK, so a second run against a different JDK is served from the cache. `--no-build-cache --rerun` avoids it.

Printing `Runtime.version()` from inside the test JVM is what caught both.

**Comment by msridhar** [comment]
Just want to check, are you running on JDK 17.0.19 or later?  That's the version that added `-XDaddTypeAnnotationsToSymbol`.  Almost certainly you are, just want to be sure.

**Comment by vlsi** [comment]
I have liberica 17.0.10 and semeru 17.0.9, so I was not running with the latest 17.

**Comment by msridhar** [comment]
Oh shoot, sorry for the confusion!  Can you update and test?

This is documented [here](https://github.com/uber/NullAway/wiki/JSpecify-Support#supported-jdk-versions).  We thought about putting more aggressive checking into NullAway around point versions but decided against, as even with 17.0.19 support is not 100% consistent (e.g., I don't think Oracle JDK supports the flag).  But, we could re-consider this, and just fail NullAway if you're in JSpecify mode and you're not running at least 17.0.19, 21.0.8, or 22+.

**Comment by vlsi** [comment]
Bisected the JDK 17 update line and narrowed the title and body accordingly: this is fixed in **17.0.19**. 17.0.18+10 fails and 17.0.19+11 passes on BellSoft Liberica, and IBM Semeru agrees, failing on 17.0.9+9 and passing on 17.0.20+8 — same Error Prone (2.42.0) throughout, so the JDK is what decides it, and two vendors moving together make it a fix in the OpenJDK 17 update line rather than a vendor build.

Two consequences worth having:

- The gate in #1789 now compares against 17.0.19 rather than skipping every JDK 17, so the bytecode variant runs wherever it works. On 17.0.19 the whole `testJdk17` suite is green with 196 bytecode runs, matching JDK 21 and the default task.
- The four `[SOURCE]` tests that `testJdk17` fails on an older JDK 17 — `WildcardTests.issue1671` and three in `BytecodeGenericsTests` — behave the same way: red on 17.0.10 and 17.0.18, green on 17.0.19. I did not investigate whether they share a cause. It does mean a local `testJdk17` failure is worth checking against the JDK patch level before anything else.

**Comment by vlsi** [comment]
Thanks — you were right, and the wiki page answers it. Updating and bisecting confirms the documented boundaries, so this is an unsupported-JDK condition rather than a NullAway defect, and it can be closed as such.

Measured with Error Prone held fixed:

| JDK | Vendor | Bytecode run |
| --- | --- | --- |
| 17.0.9+9 | IBM Semeru | fails |
| 17.0.10+13, 17.0.15+10, 17.0.17+15, 17.0.18+10 | BellSoft Liberica | fails |
| **17.0.19+11**, 17.0.20+10 | BellSoft Liberica | passes |
| 17.0.20+8 | IBM Semeru | passes |
| 21.0.2+14 | BellSoft Liberica | fails |
| **21.0.9+10** | Amazon Corretto | passes |

Two vendors move together across the 17 boundary, so it belongs to the OpenJDK update line rather than to a vendor build.

Two things I would not have found without updating:

- On a JDK 17 older than 17.0.19, `./gradlew :nullaway:testJdk17` also fails `WildcardTests.issue1671` and three tests in `BytecodeGenericsTests`. They pass from 17.0.19. Whether they share a cause with this one I did not check. CI does not see any of it, since `setup-java` with `java-version: 17` installs the latest patch.
- On 17.0.19 the whole `testJdk17` suite is green, with the same 196 bytecode runs as JDK 21 and the default task.

On failing fast in JSpecify mode: I would find that useful, and the case for it is that the failure mode today is silent and looks like a NullAway bug — the flag is accepted and ignored, so you get a diagnostic that the same code as source does not produce, with nothing pointing at the JDK. Your point about Oracle JDK not carrying the flag argues for checking the observable rather than the version: NullAway already reads `Options` for `addTypeAnnotationsToSymbol` in `JSpecifyJavacConfig.isValidJavacConfigForJSpecifyMode`, and what is missing is that an older JDK accepts `-XD` silently, so the option reads as set while nothing acts on it. If there is a way to observe the symbol rather than the flag, that would cover Oracle and the point releases with one check. Happy to look into it if you want it as a separate issue.

#1789 now gates its bytecode run on 17.0.19 / 21.0.8 / 22+ rather than skipping all of JDK 17.

**Comment by msridhar** [comment]
@vlsi I think I found a way to actually test support for the feature; I put up #1794.  When if you have time if you could take a look that'd be great.
